plugins/spender/splice.c

```diff
@@ -0,0 +1,1402 @@
+	enum splice_cmd_state old_state = state->state;
+	bool got_sigs;
+
+	tal_free(pkg);
+
+	/* DTODO: juggle serial ids correctly for cross-channel splice */
+	tok = json_get_member(buf, result, "psbt");
+	psbt = json_to_psbt(splice_cmd, buf, tok);
+
+	if (psbt_contribs_changed(splice_cmd->psbt, psbt))
+		for (size_t i = 0; i < tal_count(splice_cmd->states); i++)
+			if (splice_cmd->actions[i]->channel_id)
+				splice_cmd->states[i]->state = SPLICE_CMD_UPDATE_NEEDS_CHANGES;
+
+	assert(psbt);
+	tal_free(splice_cmd->psbt);
+	splice_cmd->psbt = tal_steal(splice_cmd, psbt);
+
+	tok = json_get_member(buf, result, "signatures_secured");
+	if (!json_to_bool(buf, tok, &got_sigs))
+		return command_fail_badparam(cmd, "signatures_secured", buf,
+					     tok, "invalid bool");
+
+	if (old_state != SPLICE_CMD_UPDATE)
+		state->state = SPLICE_CMD_UPDATE;
+	else
+		state->state = got_sigs ? SPLICE_CMD_UPDATE_DONE : SPLICE_CMD_RECVED_SIGS;
+
+	return continue_splice(splice_cmd->cmd, splice_cmd);
+}
+
+static struct command_result *splice_update(struct command *cmd,
+					    struct splice_cmd *splice_cmd,
+					    size_t index)
+{
+	struct splice_script_result *action = splice_cmd->actions[index];
+	struct out_req *req;
+	struct splice_index_pkg *pkg = tal(cmd->plugin, struct splice_index_pkg);
+
+	pkg->splice_cmd = splice_cmd;
+	pkg->index = index;
+
+	plugin_log(cmd->plugin, LOG_DBG,
+		   "splice_update(channel_id:%s)",
+		   type_to_string(tmpctx, struct channel_id,
+		   		  action->channel_id));
+
+	req = jsonrpc_request_start(cmd->plugin, cmd, "splice_update",
+				    splice_update_get_result, splice_error_pkg,
+				    pkg);
+
+	json_add_channel_id(req->js, "channel_id", action->channel_id);
+	json_add_psbt(req->js, "psbt", splice_cmd->psbt);
+
+	return send_outreq(cmd->plugin, req);
+}
+
+static struct command_result *signpsbt_get_result(struct command *cmd,
+						  const char *buf,
+						  const jsmntok_t *result,
+						  struct splice_cmd *splice_cmd)
+{
+	const jsmntok_t *tok = json_get_member(buf, result, "signed_psbt");
+	struct channel_id *channel_ids;
+
+	tal_free(splice_cmd->psbt);
+
+	splice_cmd->psbt = json_to_psbt(splice_cmd, buf, tok);
+	splice_cmd->wallet_inputs_to_signed = 0;
+
+	/* After signing we add channel_ids to the PSBT for splice_signed */
+	channel_ids = tal_arr(NULL, struct channel_id, 0);
+	for (size_t i = 0; i < tal_count(splice_cmd->actions); i++)
+		if (splice_cmd->actions[i]->channel_id)
+			tal_arr_expand(&channel_ids,
+				       *splice_cmd->actions[i]->channel_id);
+
+	psbt_set_channel_ids(splice_cmd->psbt, channel_ids);
+	tal_free(channel_ids);
+
+	return continue_splice(splice_cmd->cmd, splice_cmd);
+}
+
+static struct command_result *signpsbt(struct command *cmd,
+				       struct splice_cmd *splice_cmd)
+{
+	struct out_req *req;
+
+	req = jsonrpc_request_start(cmd->plugin, cmd, "signpsbt",
+				    signpsbt_get_result, splice_error,
+				    splice_cmd);
+
+	/* TODO: add the inputs indices we should sign */
```

Comment:
THIS IS VERY IMPORTANT, right?

Reply:
Adding to phase two splicing issue: https://github.com/ElementsProject/lightning/issues/5911

Reply:
Think this over, it is too important to move to another phase. I added input tracking to 195a73c6d0a78847b4010b56fc5fa11b30949461